### ISSUE TITLE
Hilmar Gelato Core Review - 04.11.19

### ISSUE THREAD

**Issue by hilmarx** [gelatodigital/gelato#3]
- [x] Hilmar Review 1
- [x] Luis Review and Implementation 1
- [x] further review-implement cycles?

**Comment by hilmarx** [comment]
1. ProxyRegistry / Factory / GuardFactory Setters

In what scenarios do we want to set new Factories here? Would that maybe give gelatoCore owners power which would result in a security risk to users? Not seeing one right now, but should be thought through. 

2. GelatoCoreAccounting state variables:

```
minExecutionClaimLifespan
gasOutsideGasleftChecks
gasInsideGasleftChecks 
canExecMaxGas
userProxyExecGas
```

=> Need to update the gasOutsideGasleftChecks & gasInsideGasleftChecks to reflect post Istanbul costs. This should be possible by deploying a version of gelatoCore on ropsten that uses a couple of more gasLeft()s to show how much a function consumes. 

UserProxyExecGas 100k gas looks arbitrarily set, this should be more precise. We should also reconsider canExecMaxGas to maybe be lower, depending on how many state reads a typical trigger function does. We test it with the Kyber trigger as I think ```getExpectedRate``` should be one of the most costlier public view functions, going through a lot of price feeds posted by kyber reservers. 

3. _getMinExecutionGasRequirement() has simply 4 state reads. We should be able to combine
```gasOutsideGasleftChecks``` and ```gasInsideGasleftChecks``` to minimize state reads to 3.

4. Executor Setters
I like the fact that the executor can set the price to 0, basically allowing it to charge the users through other means, such as through an action that has the executors' address hardcoded and defines a certain % of the value which was sent to be transferred to that address. This however only works if the executor can whitelist certain actions, to make sure that it is really receiving some sort of payment, which we do not support at the moment (users can mint arbitrary actions with executors), as long as they prepay, hence no executor will ever set their price to zero.

5. withdrawExecutorBalance()
=> Change to sendValue() OZ pattern by adding a require
```
(bool success, ) = recipient.call.value(amount)("");
require(success, "Address: unable to send value, recipient may have reverted");
```

6. mintExecutionClaim()
- If a user calls this func using a relayer (e.g. Gnosis Safe account), then it will be labelled as not being the ```tx.origin``` and hence will be looked at as being a proxy, which will fail. Hence the user will have to deploy a proxy with another (e.g. Gnosis safe) proxy in order to interact with gelato. This should be tested, not sure if you can do nested delegate calls.
- If the msg.sender is indeed a smart contract (tx.origin != msg.sender), then we 1) only check whether it is registered in our selected ProxyFactory, and not in our ProxyRegistry and 2) we do not check whether whether this proxy has set gelatoCore as an authority, which it must in order for ```execute()``` to work. 

Problems that could arise through that:
a) Other Proxy registries could use the same proxy factory, resulting in ```proxyFactory.isProxy``` to return true even though msg.sender is not registered in the ProxyFactory. This is a problem because front-ends check ```proxyRegistry.proxies(metaMaskAddress)``` to see if the user should be prompted with registering a proxy or not. Not sure whether this is exploitable, but it provides some uncertainties. 

b) After checking that a contract is indeed a proxy, we should also check whether gelatoCore is registered as an authority in this proxy, otherwise revert().

Regarding using tx.origin in general:

=> [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/1212) and [Vitalik](https://ethereum.stackexchange.com/questions/196/how-do-i-make-my-dapp-serenity-proof) have stated, that tx.origin should not be used. To check whether a call comes from a contract or an EOA, we can also use OZs Address library, which however also states it is not fully secure, but assuming we conduct more checks after checking if it is a contract or not, this should be fine as of now.

VB: 
> Do NOT assume that tx.origin will continue to be usable or meaningful.

Proposed changes:
```
/// @notice check if msg.sender is a user (EOA)
            if (!isContract(msg.sender)) {
                userProxy = address(proxyRegistry.proxies(msg.sender));
                require(userProxy != address(0),
                    "GelatoCore.mintExecutionClaim: msg.sender has no proxy"
                );
            } else {
                DSProxyFactory proxyFactory = DSProxyFactory(proxyFactory);
                require(proxyFactory.isProxy(msg.sender),
                    "GelatoCore.mintExecutionClaim: msg.sender is not a proxy"
                );
                userProxy = msg.sender;
                require(userProxy.authority == address(this));
            }
```

7. Why is ```address payable _userProxy``` in the ```execute() func``` payable?

8. Executor costs calcs

As we are only using the actions maxGas for making sure that the executor submitted enough gas to not run out later on at the beginning of the exec() func, we don't really need the gas calcs at the end. They are only used to emit the event with ```gasUsedEstimate``` as a gas consumption estimate, which is due to deletes inaccurate anyway and hence might simply be a waste of resources.

Another side effect is that with this implementation (not calculating payout dynamically), users always have to pay the maximum trigger gas, even if consuming only a fraction. However, I think this is ok for the first iteration. 

10. Apply sendValue to cancelExecutionClaim()

11. Problems around dealing with two types of accounts

In the mint and cancel functions, we are always dealing with 2 scenarios, one where the account owning the proxy calls the func and one where the proxy itself calls the func. Why don't we simply encode that only the proxy contracts can interact with gelato? We have done this anyway already in the Kyber use case where the proxy minted the claims for the user. For that, we would only need to create a minting script which forwards the minting requests from the user to the proxy. I don't see direct attack vectors with leaving it to both accounts to transact, however it is from a code perspective definitely cleaner to simply check whether a minter or someone who cancels is simply a proxy contract that is the claim owner or sent sufficient ether to prepay.

**Comment by hilmarx** [comment]
12. User Balances

In our current design, users simply make p2p agreement with an executor (based on trust), that the executor will execute the users transactions in the future based on an agreed upon gas price. This amount has to be prepaid by the user. The user cannot withdraw the prepaid amount, she can only cancel the claim prematurely to get the prepaid amount back.

As we pass the prepayment amount through call data, we don't really need to check ```(userProxyDeposit[_userProxy] < _mintingDeposit)```. We can even get completely rid of userproxyDeposit as a mapping.

**Comment by gitpusha** [comment]
@hilmarx  Implemented  review 1 changes - see commit referenced

**Comment by hilmarx** [comment]
> 1. Answer Luis:
on mainnet we use the Maker registry (and thereby factory) to deploy user proxies. On testnets we deploy our own registry and factory in order to mimic mainnet behavior. Hence we should check if Maker has undue power over our user proxies on mainnet. From what I saw in the code, no they dont. But we should maybe verifiy the deployed Maker registry and factory source code before we just plug into it for our use case.
Open question: should we force users to deploy a guard via us or not. I think this could potentially introduce hazards because we might block the user proxy elsewhere, where the user didnt want it to be blocked. We kinda hijack the user's proxy if he comes to us with

1. So after checking, assuming they do not have abuse powers,  is there a reason to be able to set this, other than it providing us with extra flexibility to introduce a new proxy registry? 

2. We only deploy a guard if the Authority in the users proxy is set to address(0). If not, we simply whitelist gelatoCore. If not proxy is registered in the Maker Proxy Registry, we deploy a new one and wont disturb any other services.

> _getMinExecutionGasRequirement() has simply 4 state reads. We should be able to combine
gasOutsideGasleftChecks and gasInsideGasleftChecks to minimize state reads to 3. Yep, let's do this kind of fine-tuning when our codebase is completed for launch and we have benchmarked.

We should do it before benchmarking, as otherwise, we would have to benchmark again, so lets do it now. 

> 4. Answer Luis:
> I think the idea of revenue-sharing models for executors is cool and could maybe be something once the protocol has significant usage. Certain actions could whitelist only one trusted executor e.g.
> Maybe in the future good actions will have their executor and users/frontends dont pick an executor any more. They would get the executor that the action has a deal with. This way executors would only execute actions they know and have ties with.

How would actions know which executor transacted on gelatoCore? Regarding enabling this feature I mentioned which enables 0 prepayments, we could implement it now by simply having executors whitelist action addresses with a mapping. Worth considering would be one require more and one setter more. 

> I asked the Openzeppelin guys about our particular case. They said tx.origin works and is better than their Address util in that case.
Ok nice.

> Here is my completed suggested solution - I think this solves everything - need to check for stack too deep thoo:
Looks good on first glance.

> Further need to discuss guard: should the user be in control of it or the userProxy - the former would mean potential bad claims if the user in between overwrites the guard settings. The latter could however mean the user has rigid guard settings and cannot use the proxy elsewhere anymore.

I don't quite follow and understand the issue you see with guards. The user is the owner of the proxy, the guard can be set by the owner and the proxy itself. Gelato Core interactions would break if the user uses another dapp which simply overwrites the authority to another guard contract. These front end should however check that an authority is there, and if yes, simply set themselves as another authority within it. This can of course not be enforced, but checked on the front end and providing the user with a warning if that happened (e.g. "You removed gelatos authority to transact on your behalf".

> The userProxy is address payable due to its payable fallback function (DSProxy). In some parts solc forced me to mark userProxy as address payable. I think here I do not have to do it, but I did it for clarity and consistency.

Lets remove it if it's unnecessary and we dont send value within the function. 

> yes I left it in there so we have an easy way to track these figures initially. But we could cut it out.
They, however, do not reflect the actual cost of the executor. We can keep it for testing and identifying the gas-related state variables and cut it afterwards.

> It is only slightly more complex. We can go either way. No strong opinions here as of yet.
The fewer cases we and potential auditors have to think about, the better. I am fine with both, but would recommend going with just using proxies.If someone does not have a proxy, he should not interact with gelato anyways as it wont work. 

> Yes, I think we can and should get rid of it. The hashing should keep us safe. We need to take double care of re-entrancy then though, since now a re-entrant could potentially drain all gelatoCore's ether by maliciously re-entering during cancellation for example.

You already implemented the reentrancy guard.

**Comment by hilmarx** [comment]
I guess it's time for another review!

**Comment by hilmarx** [comment]
**GelatoCore.sol:**

1) No need for re-entrancy guard in ```execute()```, as we nullify 
```hashedExecutionClaims[_executionClaimId]``` and `userProxyByExecutionClaimId[_executionClaimId]`, which, in case the action does re-enter, would lead the `execute()` to end with `return uint8(ExecutionResult.CanExecuteFailed)`  due to the return value of the ```canExecute``` call

2) No need for re-entrancy guard in `mint()` because we are not passing flow control to any address, as we are not using `transfer()` or a similar fn

3) added `require(actionGasStipend != 0, "GelatoCore.mintExecutionClaim: Action has zero gasStipend");` so that if an action has forgotten to add the gasStipend, user are protected for future failure as it already reverts here.

4) Not decided here, but I implemented the changes, you decide. Instead of reading from state 3 times the same value `userProxyExecGas.add(actionGasStipend)`, lets have the actions take care of adding the userProxyExecGas. That way user are spared extra gas costs and we have to do less calcs on core. I changed it to `actionGasStipend`, which should include the extra `21000 + smth` gas the call to the proxy should cost. See l147 & l158.

Pro: Less gas costs on core
Con: Action devs have to also think about including the proxy costs. Hence we cant say "Only measure what your actions actually costs and ignore the rest". 

5) In the `Execute()` I got rid of the unnecessary gas calcs at the end, because they dont serve any payout calculation or security purposes. They are only used for statistical purposes. However the gas values wont represent the actual consumed gas values, as they dont include the deletes. Hence, doing these calcs offchain is more accurate and less gas intensive.  l360.

6) Adding to 5, I merged the two events `LogClaimExecutedAndDeleted` and `LogExecutionResult` into one, dont need two seperate ones with two values being identical.

7) Tidied up the scope in the `cancel()` fn.

**GelatoCoreAccounting.sol**

8) Deleted re-rentrancy in `withdrawExecutorBalance`, we already nullify the executors balance. Don't need double protection. 

9) Added `require(_newMinExecutionClaimLifespan > 0);` in `setMinExecutionClaimLifespan` as self-protection for owners, as when accidently set to 0, we would break the system. 

10) Deleted everything related to `userProxyExecGas`. Again, this is debatable, I acted in the interest of gas optimisation, but would also agree to have `less action dev overhead` be rated > `gas efficiency`

**GelatoUserProxyManager.sol**

11) Don't see the point of `users.push(msg.sender)` and `userProxies.push(userProxy)`, we can easily store this data off-chain through the events. Unnecessary gas costs for users. What can be done off-chain, should be done off-chain.

12) Renamed  `_isUserProxy` to `_hasUserProxy`, makes more sense I think. We are checking if the user already has a proxy.

**GelatoUserProxy.sol** 

13) The  fn `    function setGelatoCore(address payable _gelatoCore)` is quite dangerous. One malicious action can simply change the core address an do arbitrary stuff. I would rather have us controlling the core and having it upgradable then this func in the contract. We can have a fn which maybe deactivates the core on the proxy, in case the user wants to switch off gelato?. Well, a malicious action could also simply steal the users funds directly, so this is again a debatable point. I'll leave it in for now.

**Comment by hilmarx** [comment]
See pr #8

**Comment by gitpusha** [comment]
Changes implemented (see my DONE edited-in comments above)

Will close PR and not merge due to interim code layout differences - but your suggested changes are mostly implemented in the new code: See commit #f7fbd78b25e6e009964bf1f3b18540b38ec8340e
